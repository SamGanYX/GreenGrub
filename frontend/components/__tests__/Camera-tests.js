import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';
import App from 'frontend/app/(tabs)/camera.tsx';

// Mock the necessary modules
jest.mock('expo-camera', () => ({
  CameraView: ({ children, onBarcodeScanned }) => (
    <div onClick={() => onBarcodeScanned && onBarcodeScanned({ type: 'ean13', data: '1234567890123' })}>
      {children}
    </div>
  ),
  CameraType: {
    back: 'back',
    front: 'front',
  },
  useCameraPermissions: jest.fn().mockReturnValue([{ granted: true }, jest.fn()]),
}));

jest.mock('expo-constants', () => ({
  expoConfig: {
    extra: {
      fatsecretClientId: 'test-client-id',
      fatsecretClientSecret: 'test-client-secret',
    },
  },
}));

// Mock expo-router hooks
jest.mock('expo-router', () => ({
  useNavigation: jest.fn().mockReturnValue({
    navigate: jest.fn(),
    goBack: jest.fn(),
  }),
  useFocusEffect: jest.fn().mockImplementation((callback) => callback()),
}));

global.fetch = jest.fn();
global.btoa = jest.fn().mockReturnValue('mocked-encoded-credentials');

describe('App Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly when permissions are loading', () => {
    require('expo-camera').useCameraPermissions.mockReturnValueOnce([null, jest.fn()]);
    const { toJSON } = render(
      <NavigationContainer>
        <App />
      </NavigationContainer>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when permission is not granted', () => {
    require('expo-camera').useCameraPermissions.mockReturnValueOnce([{ granted: false }, jest.fn()]);
    const { getByText } = render(
      <NavigationContainer>
        <App />
      </NavigationContainer>
    );
    expect(getByText('We need your permission to show the camera')).toBeTruthy();
  });

  it('renders camera view correctly when permission is granted', () => {
    require('expo-camera').useCameraPermissions.mockReturnValueOnce([{ granted: true }, jest.fn()]);
    const { toJSON } = render(
      <NavigationContainer>
        <App />
      </NavigationContainer>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
