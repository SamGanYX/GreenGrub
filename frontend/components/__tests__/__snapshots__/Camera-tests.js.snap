// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders camera view correctly when permission is granted 1`] = `
<View
  style={
    {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <div
    onClick={[Function]}
  >
    <View
      style={
        {
          "backgroundColor": "transparent",
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "center",
          "left": 0,
          "paddingHorizontal": 64,
          "paddingVertical": 20,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "rgba(0,0,0,0.5)",
            "borderRadius": 10,
            "opacity": 1,
            "padding": 15,
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 18,
              "fontWeight": "bold",
            }
          }
        >
          Flip Camera
        </Text>
      </View>
    </View>
  </div>
</View>
`;

exports[`App Component renders correctly when permissions are loading 1`] = `<View />`;
